rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
+
+#### Key and Value
+
+The cache implementation will reside in the `common` part of the library. In order
+to abstract away `key` and `value` entities we will need to introduce the
+following abstractions:
+
+```cpp
+namespace kernel_cache {
+// Key
+struct key_t {
+    key_t(key_impl_t *impl);
+    virtual ~key_t() = default;
+
+    bool operator==(const key_t &other) const;
+    size_t hash() const;
+private:
+    std::shared_ptr<key_impl_t> impl_;
```

Comment:
> On multi-device systems, there is a potential runtime failure mode where the impl_ structure is the same for device1 and device2, but the value_impl_t is only useable on device1. 

This would mean that the `key` is incorrect because it's supposed to be used to differentiate different values. The only way to test such cases would be using benchdnn. It is used for testing primitive cache now and if we ever decide to test oneDNN on multi-device systems then using benchdnn will be the way to do it.